src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -0,0 +1,429 @@
+      }
+    }
+    throwIfTerminal();
+    if (bufferedRecords() >= maxBufferedRecords) {
+      throw new ConnectException("Add timeout expired before buffer availability");
+    }
+    unsentRecords.addLast(record);
+    notifyAll();
+  }
+
+  /**
+   * Request a flush and block upto {@code timeoutMs} until all pending records have been flushed.
+   *
+   * If any task has failed prior to or during the flush, {@link ConnectException} will be thrown with that error.
+   */
+  public void flush(long timeoutMs) {
+    log.trace("flush {}", timeoutMs);
+    final long flushStartTimeMs = time.milliseconds();
+    try {
+      flushRequested = true;
+      synchronized (this) {
+        notifyAll();
+        for (long elapsedMs = time.milliseconds() - flushStartTimeMs;
+             !isTerminal() && elapsedMs < timeoutMs && bufferedRecords() > 0;
+             elapsedMs = time.milliseconds() - flushStartTimeMs) {
+          wait(timeoutMs - elapsedMs);
+        }
+        throwIfTerminal();
+        if (bufferedRecords() > 0) {
+          throw new ConnectException("Flush timeout expired with unflushed records: " + bufferedRecords());
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new ConnectException(e);
+    } finally {
+      flushRequested = false;
+    }
+  }
+
+  private final class BulkTask implements Callable<BulkResponse> {
+
+    final long batchId = createdBatches.incrementAndGet();
+
+    final List<R> batch;
+
+    BulkTask(List<R> batch) {
+      this.batch = batch;
+    }
+
+    @Override
+    public BulkResponse call() throws Exception {
+      final BulkResponse rsp;
+      try {
+        rsp = execute();
+      } catch (Exception e) {
+        failAndStop(e);
+        throw e;
+      }
+      log.debug("Successfully executed batch {} of {} records", batchId, batch.size());
+      onBatchCompletion(batch.size());
+      return rsp;
+    }
+
+    private BulkResponse execute() throws Exception {
+      final B bulkReq;
+      try {
+        bulkReq = bulkClient.bulkRequest(batch);
+      } catch (Exception e) {
+        log.error("Failed to create bulk request from batch {} of {} records", batchId, batch.size(), e);
+        throw e;
+      }
+      for (int remainingRetries = maxRetries; true; remainingRetries--) {
+        boolean retriable = true;
+        try {
+          log.trace("Executing batch {} of {} records", batchId, batch.size());
+          final BulkResponse bulkRsp = bulkClient.execute(bulkReq);
+          if (bulkRsp.isSucceeded()) {
+            return bulkRsp;
+          }
+          retriable = bulkRsp.isRetriable();
+          throw new ConnectException("Bulk request failed: " + bulkRsp.getErrorInfo());
+        } catch (Exception e) {
+          if (retriable && remainingRetries > 0) {
+            log.warn("Failed to execute batch {} of {} records, retrying after {} ms", batchId, batch.size(), retryBackoffMs, e);
+            time.sleep(retryBackoffMs);
```

Comment:
Should we be checking for whether we need to shut down, and/or using `wait` instead of `time.sleep`? I am wondering if the # of retries can basically exceed the amount of time we want to wait for, e.g., shutdown, in which case we'd still be trying to send data even though we really wanted to have stopped already. This obviously gets a bit tricky since you want to flush during shutdown, but maybe there's some final flag we should be setting that will force these tasks to bail out even before all retries are exhausted?